app/uk/gov/hmrc/apprenticeshiplevy/controllers/auth/AuthAction.scala

```diff
@@ -0,0 +1,240 @@
+        case _ ~ PAClientId(_) =>
+          Future.successful(Right(AuthenticatedRequest(request, None)))
+        case Enrolments(enrolments) ~ _ =>
+          val payeRef: Option[EmpRef] = enrolments.find(_.key == "IR-PAYE")
+            .flatMap { enrolment =>
+              val taxOfficeNumber = enrolment.identifiers.find(id => id.key == "TaxOfficeNumber").map(_.value)
+              val taxOfficeReference = enrolment.identifiers.find(id => id.key == "TaxOfficeReference").map(_.value)
+
+              (taxOfficeNumber, taxOfficeReference) match {
+                case (Some(number), Some(reference)) => Some(EmpRef(number, reference))
+                case _ => None
+              }
+            }
+          val isCorrectEmpRef: Boolean = !payeRef.exists(_.value != empRef.empref)
+          if(isCorrectEmpRef) {
+            Future.successful(Right(AuthenticatedRequest(request, payeRef)))
+          } else {
+            Logger.warn(s"Unauthorized request of ${empRef.empref} from $payeRef")
+            Future.successful(Left(Unauthorized(
+              Json.toJson(AuthError(UNAUTHORIZED, "UNAUTHORIZED", s"Unauthorized request of ${empRef.empref}."))
+            )))
+          }
+      }.recover { case e: Throwable => Left(ErrorHandler.authErrorHandler(e)) }
+    }
+  }
+}
+
+class PrivilegedAuthActionImpl @Inject()(val authConnector: AuthConnector)(implicit executionContext: ExecutionContext)
+  extends AuthAction with AuthorisedFunctions {
+
+  override protected def refine[A](request: Request[A]): Future[Either[Result, AuthenticatedRequest[A]]] = {
+    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers, None)
+
+    authorised(AuthProviders(PrivilegedApplication)) {
+      Future.successful(Right(AuthenticatedRequest(request, None)))
+    }.recover { case e: Throwable => Left(ErrorHandler.authErrorHandler(e)) }
+
+  }
+}
+
+@ImplementedBy(classOf[AuthActionImpl])
+trait AuthAction extends ActionBuilder[AuthenticatedRequest] with ActionRefiner[Request, AuthenticatedRequest]
+
+private object ErrorHandler {
+  private def extractReason(msg: String): String =
+    Try(if (msg.contains("Response body")) {
+      val str1 = msg.reverse.substring(1).reverse.substring(msg.indexOf("Response body") + 14).trim
+      val m = if (str1.startsWith("{")) str1 else str1.substring(str1.indexOf("{"))
+      Try((Json.parse(m) \ "reason").as[String]) getOrElse (Json.parse(m) \ "Reason").as[String]
+    } else {
+      msg
+    }) getOrElse msg
+
+  private def logWarningAboutException(e: Throwable, code: Int, description: String): Unit = {
+    val message = s"Client ${
+      MDC.get("X-Client-ID")
+    } API error: ${
+      e.getMessage
+    }, API returning $description ${
+      code
+    }"
+    Logger.warn(message)
+  }
+
+  def authErrorHandler(exc: Throwable): Result = {
```

Comment:
I'd agree this needs cleaned up. This has been carried over from old auth, so some things here may not even be relevant. We are going to be speaking with ESFA after this auth work to look at some of the stuff we send them and improve them. I think this would be the best opportunity to look at this - mainly because we are keen to get this work out so auth can decom old auth libraries